src/internal/dockertestenv/util.go

```diff
@@ -0,0 +1,125 @@
+package dockertestenv
+
+import (
+	"context"
+	"io"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	docker "github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/sirupsen/logrus"
+)
+
+func newDockerClient() docker.APIClient {
+	dclient, err := docker.NewClientWithOpts(docker.FromEnv)
+	if err != nil {
+		panic(err)
+	}
+	return dclient
+}
+
+func getDockerHost() string {
+	client := newDockerClient()
+	defer client.Close()
+	host := client.DaemonHost()
+	u, err := url.Parse(host)
+	if err != nil {
+		panic(err)
+	}
+	if u.Scheme == "unix" {
+		return "127.0.0.1"
+	}
+	return u.Hostname()
+}
+
+type containerSpec struct {
+	Image   string
+	Cmd     []string
+	PortMap map[uint16]uint16
+	Env     map[string]string
+}
+
+// ensureContainer checks that a container is running with containerName, if it is not, one will be created using image and portMap.
+// image should be the name of an image e.g. minio/minio:latest
+// portMap is a mapping from host ports to container ports.  The image/values of this mapping are taken implicitly to be the exposed container ports.
+func ensureContainer(ctx context.Context, dclient docker.APIClient, containerName string, spec containerSpec) error {
+	imageName := spec.Image
+	portMap := spec.PortMap
+	if _, err := dclient.ContainerInspect(ctx, containerName); err != nil {
+		if !isErrNoSuchContainer(err) {
+			return err
+		}
+	} else {
+		logrus.Infof("container %s exists. skip creation.", containerName)
```

Comment:
This may still need to ensure that the container is running - it may exist but be stopped.